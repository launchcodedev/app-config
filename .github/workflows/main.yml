name: main

on:
  pull_request:
  push:
      branches: '**'

jobs:
  # for branches, we only need to check Linux + Node LTS
  # note that we do linting here, but not on Windows / MacOS (no need to)
  linux-and-lts:
    runs-on: ${{ matrix.os }}

    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest, windows-latest]
        node-version: [14.x]

    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}
    - uses: bahmutov/npm-install@v1
    - run: yarn playwright install-deps chromium
    - run: yarn build
    - run: yarn build:es
    - run: yarn build:all
    - run: yarn lint
    - run: xvfb-run yarn test --coverage --testTimeout 20000
      env: { CI: 1 }
      if: matrix.os == 'ubuntu-latest'
    - run: yarn test --coverage --testTimeout 20000
      env: { CI: 1 }
      if: matrix.os != 'ubuntu-latest'

    # sanity check that plugins made a file without needing to loadConfig
    - run: yarn --cwd ./examples/rollup-project start
    - run: yarn --cwd ./examples/esbuild-project start

    - uses: codecov/codecov-action@v1
      if: matrix.os == 'ubuntu-latest'
      with:
        token: ${{ secrets.CODECOV_TOKEN }}
        files: |
          ./app-config-cli/coverage/coverage-final.json,
          ./app-config-config/coverage/coverage-final.json,
          ./app-config-core/coverage/coverage-final.json,
          ./app-config-cypress/coverage/coverage-final.json,
          ./app-config-encryption/coverage/coverage-final.json,
          ./app-config-exec/coverage/coverage-final.json,
          ./app-config-extension-utils/coverage/coverage-final.json,
          ./app-config-extensions/coverage/coverage-final.json,
          ./app-config-generate/coverage/coverage-final.json,
          ./app-config-git/coverage/coverage-final.json,
          ./app-config-inject/coverage/coverage-final.json,
          ./app-config-js/coverage/coverage-final.json,
          ./app-config-logging/coverage/coverage-final.json,
          ./app-config-main/coverage/coverage-final.json,
          ./app-config-meta/coverage/coverage-final.json,
          ./app-config-node/coverage/coverage-final.json,
          ./app-config-schema/coverage/coverage-final.json,
          ./app-config-settings/coverage/coverage-final.json,
          ./app-config-utils/coverage/coverage-final.json,
          ./app-config-v1-compat/coverage/coverage-final.json,
          ./app-config-vault/coverage/coverage-final.json,
          ./app-config-webpack/coverage/coverage-final.json,
          ./app-config-rollup/coverage/coverage-final.json,
          ./app-config-vite/coverage/coverage-final.json,
          ./app-config-electron/coverage/coverage-final.json,
        fail_ci_if_error: true
        verbose: true
