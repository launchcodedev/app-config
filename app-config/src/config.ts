import { join } from 'path';
import {
  ParsedValue,
  ParsingExtension,
  FallbackSource,
  NotFoundError,
  WasNotObject,
  ReservedKeyError,
  Json,
  isObject,
  logger,
} from '@app-config/core';
import { defaultAliases, EnvironmentAliases } from './environment';
import { FlexibleFileSource, FileSource, EnvironmentSource } from './config-source';
import { defaultExtensions, defaultEnvExtensions, markAllValuesAsSecret } from './extensions';
import { loadSchema, Options as SchemaOptions } from './schema';

export interface Options {
  directory?: string;
  fileNameBase?: string;
  secretsFileNameBase?: string;
  environmentVariableName?: string;
  extensionEnvironmentVariableNames?: string[];
  environmentOverride?: string;
  environmentAliases?: EnvironmentAliases;
  parsingExtensions?: ParsingExtension[];
  secretsFileExtensions?: ParsingExtension[];
  environmentExtensions?: ParsingExtension[];
  defaultValues?: Json;
}

export interface Configuration {
  /** full configuration plain JSON, with secrets and nonSecrets */
  fullConfig: Json;
  /** parsed configuration value, with metadata (like ConfigSource) still attached */
  parsed: ParsedValue;
  parsedSecrets?: ParsedValue;
  parsedNonSecrets?: ParsedValue;
  /** non-exhaustive list of files that were read (useful for reloading in plugins) */
  filePaths?: string[];
}

export async function loadConfig({
  directory = '.',
  fileNameBase = '.app-config',
  secretsFileNameBase = `${fileNameBase}.secrets`,
  environmentVariableName = 'APP_CONFIG',
  extensionEnvironmentVariableNames = ['APP_CONFIG_EXTEND', 'APP_CONFIG_CI'],
  environmentOverride,
  environmentAliases = defaultAliases,
  parsingExtensions = defaultExtensions(environmentAliases, environmentOverride),
  secretsFileExtensions = parsingExtensions.concat(markAllValuesAsSecret()),
  environmentExtensions = defaultEnvExtensions(),
  defaultValues,
}: Options = {}): Promise<Configuration> {
  // before trying to read .app-config files, we check for the APP_CONFIG environment variable
  const env = new EnvironmentSource(environmentVariableName);
  logger.verbose(`Trying to read ${environmentVariableName} for configuration`);

  try {
    let parsed = await env.read(environmentExtensions);

    if (defaultValues) {
      parsed = ParsedValue.merge(ParsedValue.literal(defaultValues), parsed);
    }

    verifyParsedValue(parsed);

    return { parsed, fullConfig: parsed.toJSON() };
  } catch (error) {
    // having no APP_CONFIG environment variable is normal, and should fall through to reading files
    if (!(error instanceof NotFoundError)) throw error;
  }

  logger.verbose(`Trying to read files for configuration`);

  const [mainConfig, secrets] = await Promise.all([
    new FlexibleFileSource(
      join(directory, fileNameBase),
      environmentOverride,
      environmentAliases,
    ).read(parsingExtensions),

    new FlexibleFileSource(
      join(directory, secretsFileNameBase),
      environmentOverride,
      environmentAliases,
    )
      .read(secretsFileExtensions)
      .catch((error) => {
        // NOTE: secrets are optional, so not finding them is normal
        if (error instanceof NotFoundError) {
          logger.verbose('Did not find secrets file');
          return undefined;
        }

        throw error;
      }),
  ]);

  let parsed = secrets ? ParsedValue.merge(mainConfig, secrets) : mainConfig;

  if (defaultValues) {
    parsed = ParsedValue.merge(ParsedValue.literal(defaultValues), parsed);
  }

  // the APP_CONFIG_EXTEND and APP_CONFIG_CI can "extend" the config (override it), so it's done last
  if (extensionEnvironmentVariableNames.length > 0) {
    logger.verbose(
      `Checking [${extensionEnvironmentVariableNames.join(', ')}] for configuration extension`,
    );

    const extension = new FallbackSource(
      extensionEnvironmentVariableNames.map((varName) => new EnvironmentSource(varName)),
    );

    try {
      const parsedExtension = await extension.read(environmentExtensions);

      logger.verbose(
        `Found configuration extension in $${
          parsedExtension.assertSource(EnvironmentSource).variableName
        }`,
      );

      parsed = ParsedValue.merge(parsed, parsedExtension);
    } catch (error) {
      // having no APP_CONFIG_CI environment variable is normal, and should fall through to reading files
      if (!(error instanceof NotFoundError)) throw error;
    }
  }

  const filePaths = new Set<string>();

  for (const source of mainConfig.allSources()) {
    if (source instanceof FileSource) {
      filePaths.add(source.filePath);
    }
  }

  if (secrets) {
    for (const source of secrets.allSources()) {
      if (source instanceof FileSource) {
        filePaths.add(source.filePath);
      }
    }
  }

  verifyParsedValue(parsed);

  return {
    parsed,
    parsedSecrets: secrets,
    parsedNonSecrets: mainConfig.cloneWhere((v) => !v.meta.fromSecrets),
    fullConfig: parsed.toJSON(),
    filePaths: Array.from(filePaths),
  };
}

export async function loadValidatedConfig(
  options?: Options,
  schemaOptions?: SchemaOptions,
): Promise<Configuration> {
  const [{ validate }, { fullConfig, parsed, ...rest }] = await Promise.all([
    loadSchema({
      directory: options?.directory,
      fileNameBase: options?.fileNameBase ? `${options.fileNameBase}.schema` : undefined,
      environmentOverride: options?.environmentOverride,
      environmentAliases: options?.environmentAliases,
      ...schemaOptions,
    }),
    loadConfig(options),
  ]);

  if (!isObject(fullConfig)) {
    throw new WasNotObject('Configuration was not an object');
  }

  logger.verbose('Config was loaded, validating now');
  validate(fullConfig, parsed);

  return { fullConfig, parsed, ...rest };
}

function verifyParsedValue(parsed: ParsedValue) {
  parsed.visitAll((value) => {
    for (const [key, item] of Object.entries(value.asObject() ?? {})) {
      if (key.startsWith('$') && !item.meta.fromEscapedDirective) {
        throw new ReservedKeyError(
          `Saw a '${key}' key in an object, which is a reserved key name. Please escape the '$' like '$${key}' if you intended to make this a literal object property.`,
        );
      }
    }
  });
}
