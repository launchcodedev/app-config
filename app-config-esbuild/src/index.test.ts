import { withTempFiles } from '@app-config/test-utils';
import { build } from 'esbuild';
import createPlugin from './index';

it('loads config correctly', () =>
  withTempFiles(
    {
      '.app-config.schema.yml': `
        type: object
        additionalProperties: true
      `,
      '.app-config.yml': `
        foo: bar
      `,
      'a.js': `
        import config from '@app-config/main';
        console.log(config);
      `,
    },
    async (inDir) => {
      const res = await build({
        entryPoints: [inDir('a.js')],
        plugins: [createPlugin({ loadingOptions: { directory: inDir('.') } })],
        bundle: true,
        minify: true,
        write: false,
      });

      expect(res.outputFiles[0].text).toMatchSnapshot();
    },
  ));

it('fails when config is incorrect', () =>
  withTempFiles(
    {
      '.app-config.schema.yml': `
        type: object
        additionalProperties: false
      `,
      '.app-config.yml': `
        foo: bar
      `,
      'a.js': `
        import config from '@app-config/main';
        console.log(config);
      `,
    },
    async (inDir) => {
      await expect(
        build({
          entryPoints: [inDir('a.js')],
          plugins: [createPlugin({ loadingOptions: { directory: inDir('.') } })],
          bundle: true,
          minify: true,
          write: false,
        }),
      ).rejects.toThrow(
        'error: [plugin: @app-config/esbuild] Config is invalid: config must NOT have additional properties',
      );
    },
  ));

it('loads validation function', () =>
  withTempFiles(
    {
      '.app-config.schema.yml': `
        type: object
        additionalProperties: false
        properties:
          foo: { type: string }
      `,
      '.app-config.yml': `
        foo: bar
      `,
      'a.js': `
        import { validateConfig } from '@app-config/main';

        validateConfig({ foo: 12 })
      `,
    },
    async (inDir) => {
      const res = await build({
        entryPoints: [inDir('a.js')],
        plugins: [createPlugin({ loadingOptions: { directory: inDir('.') } })],
        bundle: true,
        minify: true,
        write: false,
      });

      expect(res.outputFiles[0].text).toMatchSnapshot();
    },
  ));

it('loads currentEnvironment', () =>
  withTempFiles(
    {
      '.app-config.schema.yml': `
        type: object
        additionalProperties: false
        properties:
          foo: { type: string }
      `,
      '.app-config.yml': `
        foo: bar
      `,
      'a.js': `
        import { currentEnvironment } from '@app-config/main';

        console.log(currentEnvironment())
      `,
    },
    async (inDir) => {
      const res = await build({
        entryPoints: [inDir('a.js')],
        plugins: [createPlugin({ loadingOptions: { directory: inDir('.') } })],
        bundle: true,
        minify: true,
        write: false,
      });

      expect(res.outputFiles[0].text).toMatchSnapshot();
    },
  ));
