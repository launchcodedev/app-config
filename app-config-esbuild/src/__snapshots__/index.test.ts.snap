// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads config correctly 1`] = `
"(()=>{var r={foo:\\"bar\\"},e=(typeof window==\\"undefined\\"?globalThis:window)||{},t=e._appConfig||r;(typeof e._appConfig==\\"undefined\\"||!Object.isFrozen(e._appConfig))&&(e._appConfig=t);var a=t;console.log(a);})();
"
`;

exports[`loads currentEnvironment 1`] = `
"(()=>{var o={foo:\\"bar\\"},e=(typeof window==\\"undefined\\"?globalThis:window)||{},r=e._appConfig||o;(typeof e._appConfig==\\"undefined\\"||!Object.isFrozen(e._appConfig))&&(e._appConfig=r);function t(){return\\"test\\"}console.log(t());})();
"
`;

exports[`loads validation function 1`] = `
"(()=>{var h={foo:\\"bar\\"},n=(typeof window==\\"undefined\\"?globalThis:window)||{},v=n._appConfig||h;(typeof n._appConfig==\\"undefined\\"||!Object.isFrozen(n._appConfig))&&(n._appConfig=v);function P(){let s={};return function(p){\\"use strict\\";p.exports=i,p.exports.default=i;var b={type:\\"object\\",additionalProperties:!1,properties:{foo:{type:\\"string\\"}},$schema:\\"http://json-schema.org/draft-07/schema#\\"};function i(t,o){\\"use strict\\";if(o)var a=o.instancePath,g=o.parentData,m=o.parentDataProperty,y=o.rootData;else var a=\\"\\",g=void 0,m=void 0,y=t;var e=null,r=0;if(t&&typeof t==\\"object\\"&&!Array.isArray(t)){for(var f in t)if(f!==\\"foo\\"){var c={instancePath:a,schemaPath:\\"#/additionalProperties\\",keyword:\\"additionalProperties\\",params:{additionalProperty:f},message:\\"must NOT have additional properties\\"};e===null?e=[c]:e.push(c),r++}if(t.foo!==void 0&&typeof t.foo!=\\"string\\"){var u={instancePath:a+\\"/foo\\",schemaPath:\\"#/properties/foo/type\\",keyword:\\"type\\",params:{type:\\"string\\"},message:\\"must be string\\"};e===null?e=[u]:e.push(u),r++}}else{var d={instancePath:a,schemaPath:\\"#/type\\",keyword:\\"type\\",params:{type:\\"object\\"},message:\\"must be object\\"};e===null?e=[d]:e.push(d),r++}return i.errors=e,r===0}}(s),s.exports}var l=P();l({foo:12});})();
"
`;
